// @lexical/code-core/LexicalCodeCore.node.mjs
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const mod = await (process.env.NODE_ENV !== 'production' ? import('./LexicalCodeCore.dev.mjs') : import('./LexicalCodeCore.prod.mjs'));
export const $createCodeHighlightNode = mod.$createCodeHighlightNode;
export const $createCodeNode = mod.$createCodeNode;
export const $getCodeLineDirection = mod.$getCodeLineDirection;
export const $getEndOfCodeInLine = mod.$getEndOfCodeInLine;
export const $getFirstCodeNodeOfLine = mod.$getFirstCodeNodeOfLine;
export const $getLastCodeNodeOfLine = mod.$getLastCodeNodeOfLine;
export const $getStartOfCodeInLine = mod.$getStartOfCodeInLine;
export const $isCodeHighlightNode = mod.$isCodeHighlightNode;
export const $isCodeNode = mod.$isCodeNode;
export const CodeExtension = mod.CodeExtension;
export const CodeHighlightNode = mod.CodeHighlightNode;
export const CodeNode = mod.CodeNode;
export const DEFAULT_CODE_LANGUAGE = mod.DEFAULT_CODE_LANGUAGE;
export const getDefaultCodeLanguage = mod.getDefaultCodeLanguage;
